### ISSUE TITLE
Does the specification intend for a null ClassLoader to mean something specific?

### ISSUE THREAD

**Issue by ljnelson** [eclipse/microprofile-config#426]
The [javadocs for `ConfigProvider#getConfig(ClassLoader)`](https://static.javadoc.io/org.eclipse.microprofile.config/microprofile-config-api/1.3/org/eclipse/microprofile/config/ConfigProvider.html#getConfig-java.lang.ClassLoader-) say, in part:

> Provide a `Config` based on all `ConfigSources` of the specified `ClassLoader`
> There is exactly a single Config instance per ClassLoader

The [javadocs for `ConfigBuilder#forClassLoader(ClassLoader)`](https://static.javadoc.io/org.eclipse.microprofile.config/microprofile-config-api/1.3/org/eclipse/microprofile/config/spi/ConfigBuilder.html#forClassLoader-java.lang.ClassLoader-) say, in part:

> Return the ConfigBuilder for a given classloader
> **Parameters:**
> `loader - the specified classloader`
> **Returns:**
> `the ConfigureBuilder` [_sic_] `for the given classloader`

Neither of these javadocs says whether a `null` `ClassLoader` supplied here is legal, and, if so, what an implementation must do.

Then the [javadocs for `ConfigProviderResolver#registerConfig(Config, ClassLoader)`](https://static.javadoc.io/org.eclipse.microprofile.config/microprofile-config-api/1.3/org/eclipse/microprofile/config/spi/ConfigProviderResolver.html#registerConfig-org.eclipse.microprofile.config.Config-java.lang.ClassLoader-) say, in part:

> Register a given `Config` within the Application (or Module) identified by the given ClassLoader. If the ClassLoader is `null` then the current Application will be used.
> **Parameters:**
> `config - which should get registered`
> `classLoader - which identifies the Application or Module the given Config should get associated with.`

This sort of suggests that if the supplied `ClassLoader` is `null` then the current thread's context classloader should be used instead.

When `ClassLoader`s are `null` in all of these cases, must they all behave as if a non-`null` `ClassLoader` were supplied?  If so, is the current thread's context classloader the proper one to use?  If not, what if any requirements are placed on implementations in each case?

**Comment by Emily-Jiang** [comment]
> The javadocs for ConfigProvider#getConfig(ClassLoader) say, in part:

    Provide a Config based on all ConfigSources of the specified ClassLoader
    There is exactly a single Config instance per ClassLoader

The javadocs for ConfigBuilder#forClassLoader(ClassLoader) say, in part:

    Return the ConfigBuilder for a given classloader
    Parameters:
    loader - the specified classloader
    Returns:
    the ConfigureBuilder [sic] for the given classloader

Neither of these javadocs says whether a null ClassLoader supplied here is legal, and, if so, what an implementation must do.

Then the javadocs for ConfigProviderResolver#registerConfig(Config, ClassLoader) say, in part:

    Register a given Config within the Application (or Module) identified by the given ClassLoader. If the ClassLoader is null then the current Application will be used.
    Parameters:
    config - which should get registered
    classLoader - which identifies the Application or Module the given Config should get associated with.

This sort of suggests that if the supplied ClassLoader is null then the current thread's context classloader should be used instead.

When ClassLoaders are null in all of these cases, must they all behave as if a non-null ClassLoader were supplied? If so, is the current thread's context classloader the proper one to use? If not, what if any requirements are placed on implementations in each case?

**Comment by ljnelson** [comment]
> The classloader in this circumstances are not null.

So the javadocs for `ConfigBuilder#forClassLoader(ClassLoader)` should indicate that `null` must not be supplied, correct?  And should specify that a `NullPointerException` will be thrown?  And there should be a TCK test for this?  Or…?  Just trying to understand the contract.

**Comment by dmlloyd** [comment]
This is all further complicated by the fact that the JVM may use `null` as a special class loader value in some cases.  For example:

```console
$ echo "String.class.getClassLoader()" | jshell
|  Welcome to JShell -- Version 11.0.3
|  For an introduction type: /help intro

jshell> String.class.getClassLoader()
$1 ==> null
```

So, `null` also means "a bootstrap class loader" in some contexts.  Confusing!

My interpretation of the rules is essentially thus: passing in a `null` class loader to any of the above methods will result in first checking the TCCL, and if the TCCL is `null`, using the system class loader.  Thus `null` is a meaningful value by this interpretation, and allows for cases where some or all classes have a class loader of `null`.

Implementation-wise it's easy enough to create a "stand in" for the `null` system class loader like this snippet that I've copied from my working tree of SmallRye Config:

```java
    private static ClassLoader calculateSystemClassLoader() {
        ClassLoader cl = ClassLoader.getSystemClassLoader();
        if (cl == null) {
            // non-null ref that delegates to the system
            cl = new ClassLoader(null) {
            };
        }
        return cl;
    }
```

So if there is agreement on this interpretation, I'd recommend that the language be amended in all locations to be something like this:

```java
/**
 * (...)
 * @param classLoader the class loader to use; if the value is {@code null}, then the thread's {@linkplain Thread#getContextClassLoader() context class loader} is used; if that value is {@code null}, then the system class loader is used
 * (...)
 */
```

**Comment by ljnelson** [comment]
I don't want to throw this off the rails, but given that the classloader is not actually used is there any particular (historical?) reason this registration facility is used, and, if so, why a `ClassLoader` has been chosen to be the key?  Couldn't it just be `Object`?

For completeness, I understand in an application server environment there might be multiple applications, and clearly for non-CDI cases you have to reinvent CDI scopes (partially), but surely the burden of uniqueness could be foisted off on the caller?  Just make it a non-`null` `Object`.

(Of course if you're in a MicroProfile environment you are required to have CDI, so surely all of this could just be implemented on top of `BeanManager.getReference()`, for those who didn't want to use injection, and then you could actually _use_ CDI scopes instead of reinventing them.  Shoot! I'm heading off the rails!)

**Comment by dmlloyd** [comment]
Yes, I see what you're saying, though it starts to become a question of "how should we redesign this specification" rather than "can we refine the existing specification to take off the rough edges" at some point.

**Comment by ljnelson** [comment]
Right, but start by changing `ClassLoader` to `Object` and then do away with any language that might exist (I've forgotten) about what happens when a `ClassLoader` goes out of scope.  Backwards-compatible change that opens the door to much easier implementations IMHO.

**Comment by dmlloyd** [comment]
The change would not be backwards compatible from either a source or binary perspective AFAICT.  You'd be changing the parameter type of methods that can be both called and implemented by the user.

**Comment by ljnelson** [comment]
Maybe I used the wrong term.  Code that calls the method today and supplies it with a `ClassLoader` (the only possible calls today) would continue to work, unchanged, if the parameter type changed to `Object`…?

**Comment by dmlloyd** [comment]
It would not; such code would get a `NoSuchMethodError` at runtime unless it was recompiled against the updated specification.  This is binary incompatibility: compiled methods in Java are encoded with complete signature information including method parameter *and* return types, and callers must exactly match callees to be binary compatible.

Of course classes which implement the old version would break entirely, and would not, as I said, be source *or* binary compatible.  So just adding a "bridge" that matches the old method signature doesn't work either as a solution.

**Comment by ljnelson** [comment]
Ah! You learn something new (in this case, for me, absolutely earth-shaking, upending everything I thought I knew about polymorphism) every day!  Off to write test cases and explore this!

**Comment by ljnelson** [comment]
For posterity, it appears that MicroProfile Config [may be introducing breaking changes soon anyway](https://groups.google.com/d/msg/microprofile/bkHtO_JhB9M/kt8vwANcBwAJ) so perhaps some of the spec repairs mentioned above could be considered then.

**Comment by Emily-Jiang** [comment]
Can this issue be closed?

**Comment by ljnelson** [comment]
Was there a definitive answer that refers to the specification?

**Comment by dmlloyd** [comment]
I think we'll need to update the specification.  I've added it to my todo list.

**Comment by dmlloyd** [comment]
This issue is not resolved IMO.  I clarified the language but the actual normative specification was not necessarily sufficiently clarified.  Stop closing issues before they're verified!

**Comment by dmlloyd** [comment]
😦

**Comment by ljnelson** [comment]
Given that [my comment](https://github.com/eclipse/microprofile-config/issues/426#issuecomment-519975659) was never answered I would vote for this issue to remain open until an answer is provided.

**Comment by dmlloyd** [comment]
The issue should remain open until there is consensus (at least among committers) as to the resolution.

**Comment by Emily-Jiang** [comment]
Trying to tidy up this issue:

> 

    The classloader in this circumstances are not null.

So the javadocs for ConfigBuilder#forClassLoader(ClassLoader) should indicate that null must not be supplied, correct? And should specify that a NullPointerException will be thrown? And there should be a TCK test for this? Or…? Just trying to understand the contract.

The remaining action is to add Spec clarification to disable null to be passed in with a TCK test for that.

**Comment by ljnelson** [comment]
I just ran into an issue today that brought this back into the spotlight.

Because of [WELD-2494](https://issues.redhat.com/browse/WELD-2494), which was in reaction to [JDK-8172726](https://bugs.openjdk.java.net/browse/JDK-8172726), there are cases during CDI container initialization where the context classloader is `null`.  Basically, use `ConfigProvider#getConfig()` in a portable extension during bean deployment in CDI 2.0/SE and, since by default in the SE flavor of Weld, the common fork/join pool is used unless you take steps to ensure otherwise and therefore WELD-2494 is in effect, you'll see that the context classloader is `null`.  Pretty reliable.

Should the system classloader _really_ be used in such situations, or should some other `ClassLoader` be used instead?  Obviously the specification currently says nothing here but maybe it should?

**Comment by radcortez** [comment]
If I understood correctly, we just need to clarify that if the context classloader is null then we fallback to the System classloader? If not, what other classloader should be used? Do we have any other alternative? What was your idea?

**Comment by ljnelson** [comment]
The classloader of `Config` itself?  Something that isn't variable or left up to the VM or the environment?

**Comment by radcortez** [comment]
Using the `Config` classloader makes sense.
